### ISSUE TITLE
High CPU consumption on Samsung in v4.6.4

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#5505]
### Please answer the following questions for yourself before submitting an issue

- [X] Filters were updated before reproducing an issue
- [X] I checked the [knowledge base](https://adguard.com/kb/) and found no answer
- [X] I checked to make sure that this issue has not already been filed

### AdGuard version

v4.6.4 (204)

### Environment

* brand = samsung
* manufacturer = samsung
* model = SM-A256B
* version = 14
* security_patch = 2024-10-01

### HTTPS filtering

- [X] yes, I do

### Root access

- [ ] yes, I have it

### Integration with AdGuard VPN

- [ ] yes, I do

### Routing mode

Local VPN

### Which DNS server do you use?

DNS protection disabled

### DNS protocol

None

### Issue Details

The issue was originally reported by @derKief, check out the logs analysis here:
https://github.com/AdguardTeam/AdguardForAndroid/issues/5499#issuecomment-2460612039

To sum it up, AdGuard v4.6 consumes several times more than v4.5

### Actual Behavior

The usage should be more or less the same in v4.5 and v4.6.

**Comment by ameshkov** [comment]
Things to check:

1. Is it specific to Samsung?
2. One thing that's unusual compared to regular configuration is that DNS protection is disabled and Android's private DNS is used instead.

**Comment by derKief** [comment]
@ameshkov 

- Is it specific to Samsung? **_My other device that has the same symptoms is also a Samsung._** 

- One thing that's unusual compared to regular configuration is that DNS protection is disabled and Android's private DNS is used instead. **_I changed it during the whole troubleshooting because I suspected it, but it makes no difference._**

**Comment by 0xrxL** [comment]
I don't know about 4.6, but this is the last stats from Nightly 44:

![Screenshot_20241107_205512_AdGuard](https://github.com/user-attachments/assets/926f3798-2606-4db3-8f6c-ba2f2e95886d)
![Screenshot_20241107_210350_Device care](https://github.com/user-attachments/assets/4cbd73c5-30ef-4d49-95f8-a086442f74c8)


and CPU foreground usage is weirdly high, despite system battery usage is 1.6%.

**Comment by 0xrxL** [comment]
@ameshkov Can this usage related to this issue in any way?

**Comment by ameshkov** [comment]
@0xrxL if this is a 24-hours usage then it looks good to me.

@derKief on the contrary faces a real issue. We've been trying to reproduce it on Samsung devices yesterday but no success so far.

**Comment by ameshkov** [comment]
@derKief just in case, could you please try nighlty version of AG? It has just minor improvements, but for some reason we don't see reports from Samsung owners on the nightly build.

**Comment by abryant-hv** [comment]
Ever since at least 4.6.4, Adguard has been the top consumer of battery life on my (Samsung) phone every time I've checked, with no change in my usage of the phone. I'll try the nightly today, but 4.6.204 definitely did not fix the issues. And yes, I have sent the logs to support, but I got an email this morning closing the ticket, with no other information. 

I don't want to just pile on this ticket with a useless comment, but this issue is definitely not resolved.

**Comment by PavelParkhomenko** [comment]
> And yes, I have sent the logs to support, but I got an email this morning closing the ticket, with no other information.
> 
We responded to you yesterday in ticket #978524 before closing it. In our response, we informed you that the app logs were forwarded to our development teams and provided a link to this related issue so you can track further progress.

**Comment by abryant-hv** [comment]
Yes there was a link to this issue, which is why I'm here. It sounded above like victory was being declared a bit too early, but perhaps I misread that. 

Back on topic, I've used the v4.7 Nightly 44 this morning for about 3 hours, and AdGuard has used 19.1% of my battery in that time, 3 times more than the next app. Device battery usage is still higher than normal with the latest nightly. :(

**Comment by ameshkov** [comment]
@abryant-hv what device are you using just in case?

**Comment by abryant-hv** [comment]
It's a "Galaxy A52 5G". Model name is "SM-A526U"

**Comment by derKief** [comment]
with the Nightly 44 (4.7.89) it has gotten better but in direct comparison to the stable 4.5 it is still too high/much.

@ameshkov  I answered your private e-mail to me. Waiting on feedback.

**Comment by yazhamuthan-aruchamy** [comment]
Hi, I am using 
Device : Samsung Galaxy A35 5G.
Model Name : SM-A356E/DS.
I just updated to stable version 4.6.4
![Screenshot_20241109_074114_AdGuard](https://github.com/user-attachments/assets/df7778e8-1bb9-4cbe-8c6d-67306dab3829)
Also the last 24 Hours battery for me is Is normal.
![Screenshot_20241109_074058_AdGuard](https://github.com/user-attachments/assets/7a225d5b-64c4-4231-81df-e885319e1753)
For the past few days the usage was normal to me and I Did not see any thing abnormal
![Uploading Screenshot_20241109_074114_AdGuard.jpg…]()

https://github.com/user-attachments/assets/8b91c5de-14f1-4c4f-a3c7-19a965bb6b4c

Can you please see this because i have checked in my realms X2 same result so I think there is no problem in here. IF this is a problem please tell me. Thanks

**Comment by derKief** [comment]
Today I tried to install/update to the new 4.7 Nightly 45, but it didn't work. The app crashed and wouldn't start anymore. 
Since I don't really have the time to do any more error analysis, I've gone back to the stable version 4.5, which works 
without any problems. Hopefully this is fixed soon ....

**Comment by 0xrxL** [comment]
> Today I tried to update to the new 4.7 Nightly 45, but it didn't work. The app crashed and wouldn't start anymore. Since I don't really have the time to do any more error analysis, I've gone back to the stable version 4.5, which works without any problems. Hopefully this is fixed soon ....

Mhh...weirdly in my case the Nightly 45 (at least for the moment) seems to be very optimised. Check it out here:

![Screenshot_20241109_173142_Device care](https://github.com/user-attachments/assets/9808cdf2-d306-4530-805b-1d586d938806)

**Comment by Cb1231ct** [comment]
Had this been fixed yet. I had to turn it off the drain was so bad

**Comment by jordansworld** [comment]
no. i would just go back to 4.5 for now

**Comment by Cb1231ct** [comment]
Yes where is it to download please

**Comment by jordansworld** [comment]
https://www.apkmirror.com/apk/adguard-software-limited/adguard/adguard-4-5-25-release/

**Comment by derKief** [comment]
or directly from here -> https://github.com/AdguardTeam/AdguardForAndroid/releases/tag/v4.5

**Comment by solkarnar** [comment]
I know it's a Samsung ticket but i have the same problem on a Xiaomi Phone. I have reverted to 4.5 and will be watching this ticket to see if any solutions manifests itself.

**Comment by jordansworld** [comment]
good to know its an all around issue

**Comment by derKief** [comment]
> I know it's a Samsung ticket but i have the same problem on a Xiaomi Phone. I have reverted to 4.5 and will be watching this ticket to see if any solutions manifests itself.

I can now confirm that this is not just a Samsung problem. My wife's Poco X6 Pro is also affected and here too we are back to v4.5

**Comment by 0xrxL** [comment]
Unfortunately, today the latest Nightly started to drain my battery again, and this after starting to watch a TV series from web browser. This problem didn't happen during normal navigation, maybe this can helpful to find the root of problem @ameshkov?

**Comment by unknown4849** [comment]
Xiaomi (poco F6) too. i first think my devices broken XD. after restart and start adguard seem fine.

security patch same 2024-10 -1, could this be the cause? i feel before update to this patch don't have this problem, but after update at today i got this problem CPU locked at high speed (via CPUZ)

**Comment by ameshkov** [comment]
@0xrxL why do you think it's a "drain" and not a normal usage? Did it stop after you finished watching? Is it specific to that website?

**Comment by ameshkov** [comment]
@muchqs tbh temperature numbers does not provide us with any insights into the issue.

AdGuard prints CPU usage numbers measured by the system, it'd be helpful if you compare them between different versions.

Basically, we need every record in the log that contains `Calculated battery usage` (it's printed to the normal INFO log as well). I explained them here: https://github.com/AdguardTeam/AdguardForAndroid/issues/5499#issuecomment-2460612039

**Comment by 0xrxL** [comment]
> @0xrxL why do you think it's a "drain" and not a normal usage?

Because the system usage jumped from 1.1% to 6.4% before the end of first episode (of 50 minutes), and it's not replicable with 4.4.1/4.5 version.

> Did it stop after you finished watching? Is it specific to that website?

Yes. Try with MixDrop host.

**Comment by jordansworld** [comment]
is the version you are talking about nightly 46 or is it not released yet? @ameshkov

**Comment by ameshkov** [comment]
@jordansworld you can try nightly, but it does not yet extra logging necessary for troubleshooting.

**Comment by jordansworld** [comment]
ok i will wait

**Comment by scy0n** [comment]
Reading through this case and #5499, every model I've spot checked has an Exynos chip or is Exynos derived (e.g. Google Tensor), is anyone seeing excessive battery drain issues on 4.6 while using Snapdragon CPUs?

**Comment by jordansworld** [comment]
> Reading through this case and #5499, every model I've spot checked has an Exynos chip or is Exynos derived (e.g. Google Tensor), is anyone seeing excessive battery drain issues on 4.6 while using Snapdragon CPUs?

yes im on galaxy s23 ultra

**Comment by derKief** [comment]
> Reading through this case and #5499, every model I've spot checked has an Exynos chip or is Exynos derived (e.g. Google Tensor), is anyone seeing excessive battery drain issues on 4.6 while using Snapdragon CPUs?

My wifes Poco X6 Pro is using a MediaTek Dimensity 8300-Ultra
So it probably has nothing to do with the CPU model...

**Comment by Guguloi** [comment]
@DonEstefan. Strange. My Pixel 8 Pro is completely fine with 4.6.204.

![adguard46204](https://github.com/user-attachments/assets/ede26c67-8d6a-407c-910e-d3bc28c5cd25)

**Comment by solkarnar** [comment]
Can you enlighten us why you guys insist on having an app version different from probably your build number. I see a new version 4.6.5 and i get greeted by a 4.6.207 when i click what's new. It confuses the non tech user a lot.

Why isn't 4.6.5 (build 207) or 4.6.5.207 a thing?

**Comment by 0xrxL** [comment]
If you have installed the new version with the additional logs, try to see some movies from MixDrop, Netflix...or at least a service that offers a continuous video playback. In this way the battery drain bug shows up right away.

**Comment by derKief** [comment]
@ameshkov 
I have created the extended logs and sent them to you. Unfortunately, nothing has changed for me with the new version.
The battery consumption is still high and the device heats up more than normal -> Galaxy A25
For now i will switch back to stable v4.5

**Comment by ameshkov** [comment]
@derKief yay, thank you so much!

The problem was identified and it's in the filtering engine, the TCP/IP stack implementation consumes 100% CPU and we're already looking into what change caused this.

**Comment by ameshkov** [comment]
> i get greeted by a 4.6.207 when i click what's new

This is a bug indeed, we'll fix this.

**Comment by jordansworld** [comment]
> @derKief yay, thank you so much!
> 
> The problem was identified and it's in the filtering engine, the TCP/IP stack implementation consumes 100% CPU and we're already looking into what change caused this.

so good to hear! im so happy. thankyou!

**Comment by solkarnar** [comment]
Very glad to hear the bug has been identified! Thanks!

**Comment by jordansworld** [comment]
will this be released before the weekend? @ameshkov 😁

**Comment by ameshkov** [comment]
@jordansworld well, if we're lucky to figure out what part of the code is triggering this then yes:)

**Comment by jordansworld** [comment]
thankyou for your hard work. 🥰

**Comment by 0xrxL** [comment]
> @derKief yay, thank you so much!
> 
> The problem was identified and it's in the filtering engine, the TCP/IP stack implementation consumes 100% CPU and we're already looking into what change caused this.

FINALLY!👍This is a very good news.

**Comment by ameshkov** [comment]
Yeah, but that's still not all since we still cannot reproduce it on our own. Currently, we are very carefully reviewing the changes that were made to TCP/IP stack. If we can't notice a mistake this way, we'll make a new build with extra logging in TCP/IP stack.

**Comment by ameshkov** [comment]
Careful reading is not enough to figure it out :(

We'll prepare an APK with extra logging related to TCP/IP stack, hopefully it will help.
This APK is not for everyday use though, I'll post it here when it's ready.

**Comment by jordansworld** [comment]
i am sad. really hope this can be fixed

**Comment by jordansworld** [comment]
@ameshkov any news

**Comment by 0xrxL** [comment]
Tagged as fixed...why?

**Comment by ameshkov** [comment]
@0xrxL it's automatic, there's a merged pull request (a proposal to make changes to the code) linked to this issue.

Usually, it means that the task is fixed, but not in this case, this pull request is the one that prepares the test build with additional logging.

Besides that we identified one suspicious place in the code that could in theory be responsible for the bug, but I wouldn't bet on that since it was present in older versions too.

**Comment by ameshkov** [comment]
We just published a new version to the nightly channel.

What's inside:
1. CoreLibs update with one improvement to TCP/IP stack that can potentially help.
2. However, I am not too optimistic about that so it also writes more information to the debug log that will help us figure it out if the change doesn't help.

@derKief could you please re-do the [log collection](https://github.com/AdguardTeam/AdguardForAndroid/issues/5505#issuecomment-2471045627) as you already did once?

**Comment by ameshkov** [comment]
@DonEstefan could you please export logcat then? `adb logcat > logcat.log`

**Comment by derKief** [comment]
@ameshkov
I just sent you an email with the current logs.
No improvement in my case ... exactly the same symptoms.
I switch back to v4.5 now.

Hopefully the problem can be solved soon.

**Comment by ameshkov** [comment]
@derKief I haven't received the email yet :(

**Comment by derKief** [comment]
> @derKief I haven't received the email yet :(

Blocked by Google 
This message was blocked because its content presents a potential

   552-5.7.0 security issue. To review our message content and attachment

   content 552-5.7.0 guidelines, go to 552 5.7.0

   https://support.google.com/mail/?p=BlockedMessage`

**Comment by ameshkov** [comment]
Arghhh, Google what're you doing to me:(

I'll send you my backup email address in private.

**Comment by ameshkov** [comment]
Yay, received it!

**Comment by 0xrxL** [comment]
Wow...fixing this bug is turning into a mythological story. 😄

**Comment by ameshkov** [comment]
It is pretty mystic indeed.

Here's what we understood from @derKief's log:
1. Several TCP packets in one connection get stuck.
2. Still have no idea why, but maybe if we learn what app causes this we'll be able to reproduce it (unfortunately, this info is missing from the log that we received, need one more, sorry).

**Comment by solkarnar** [comment]
You should put out this story on your blog after you resolve it

**Comment by derKief** [comment]
At Andrey's request, I tried to create another log today and had installed the Nightly build 52 for this... and what can I say, I can't reproduce the problem anymore.
Can anyone confirm this?
I'll be monitoring this over a longer period of time

**Comment by 0xrxL** [comment]
@derKief I think you should tag users to get their response.

**Comment by derKief** [comment]
Hi Andrey, I've had the whole thing running for 24 hours now and the problems with the CPU cores/frequencies and the high temperature have not reoccurred. 
The battery consumption has also normalized somewhat but is still higher in direct comparison with stable version 4.5. 
What is noticeable here is the higher CPU background activity inside the Adguard battery stats.

My summary: the problem has been significantly improved but is still not as good as it was in version 4.5.

**Comment by ameshkov** [comment]
@derKief even though it's not as bad as it was, could you please enable debug-level logging and export it after some long using of the app? Maybe this "packet getting stuck" issue still happening but for shorter periods of time?

@DonEstefan in your case it's related to Firewall not being able to compose a notification for Google Health app:
`11-21 18:31:38.159 19186 19387 D C.V     : 18:31:38.159 [firewall-manager-notification-rules-pool-28-thread-1] No UID associated with packageName com.google.android.apps.healthdata, cannot assemble firewall notification rules`

We're looking into it.

**Comment by derKief** [comment]
I won't have the opportunity to test for a while now and have therefore gone back to the stable version 4.5. 
The troubleshooting will have to take place without me for now. Sorry

**Comment by 0xGingi** [comment]
On Nightly 54, after a full days usage, reported battery issue is only at 1%, think it's been fixed

**Comment by solkarnar** [comment]
Any idea when nightly 54 will be pushed to production?

**Comment by 0xrxL** [comment]
> Any idea when nightly 54 will be pushed to production?

So...nightly 54 fix battery drain?

I'm asking this because I can't reproduce battery drain too...for the moment.

**Comment by derKief** [comment]
The problem is not solved for everyone. In my case, the problem with the high CPU load/frequencies and the resulting high temperatures was solved. However, the increased battery consumption is still there. You can see this quite well from the high CPU background activity in the Adguard statistics. I was able to compare this quite well with my S24, which runs on v4.5.

Its better but not fixed (at least for me). I already mention this [here](https://github.com/AdguardTeam/AdguardForAndroid/issues/5505#issuecomment-2491025604)

**Comment by ameshkov** [comment]
@DonEstefan regarding the issue that you sent we figured it out. The problem was caused by a completely different bug: when you switch "firewall notifications" it was doubling the number of firewall rules (you can't see them though). In your case the number of firewall rules is probably through the roof and this is causing everything else (including the logs export issue). In the next nightly update we'll remove those duplicate rules, but you can quickly fix it too just by clearing app data.

@derKief I honestly don't understand what could've fixed that high temperature issue. I am of course glad that it's gone, but I suspect something is still wrong. If at some point you'll have time to collect the logs again, please let me know.

**Comment by AlgorithmOfGithub** [comment]
@ameshkov Hi, i have Samsung S24+, for me  not drain battery but when i close the app notifications, the app's battery usage decrease and since from 4.5 when i close notifications from app settings, the app going to crash and i have to restart app again.  Maybe gonna help for the issue.

**Comment by ameshkov** [comment]
> when i close notifications from app settings

Could you please explain what exactly you're doing here? Maybe a screenshot.

**Comment by AlgorithmOfGithub** [comment]
> > when i close notifications from app settings
> 
> Could you please explain what exactly you're doing here? Maybe with a screenshot?

step by step

- download ag v4.7 n57
- install default (licence, https filtering, cert, vpn, etc.)
- go to phone settings (for samsung) > apps > Adguard > notifications > remove tick from 'allow notifications'
- [vpn logo out] go to adguard > just showing adguard symbol

**Comment by Versty** [comment]
@AlgorithmOfGithub We have managed to reproduce this bug on our side. 
Will be fixed, you can track progress in the https://github.com/AdguardTeam/AdguardForAndroid/issues/5523 issue.

**Comment by jordansworld** [comment]
i know this hasn't been fixed even. but my battery ussge today shows ADGuard used 4 percent only
![Screenshot_20241202_153254_Device care](https://github.com/user-attachments/assets/fee4d0af-fd83-4d2a-95ae-39284359847a)

**Comment by jordansworld** [comment]
![Screenshot_20241202_153607_Device care](https://github.com/user-attachments/assets/52f9f5ef-a502-460f-947e-07898dcf8bdb)

for a 7 hour period only 4 percent is good wouldn't you say? @ameshkov

**Comment by derKief** [comment]
@ameshkov
were my last logs that I sent you a few days ago able to help in any way?

**Comment by 0xrxL** [comment]
> @ameshkov were my last logs that I sent you a few days ago able to help in any way?

Seems that log from another another user was helpful to find the bug, in fact CoreLibs will be updated soon with the fix. ☝️

**Comment by derKief** [comment]
Version 4.7 stable has been released. There is nothing about the topic in the changelog. 
I'm a little surprised that there hasn't been any feedback from the developers for a few days.

**Comment by derKief** [comment]
With the stable version 4.7 I have the problem again that the CPU cores are suddenly running at maximum frequency again. The device is also getting noticeably warm again. So a quick feedback from me: the stable version 4.7 is just as bad as the stable version 4.6. Some 4.7 nightlies that I tested did it better. I don't understand it and hope that a solution will finally be found. The constant testing is starting to annoy me.

**Comment by 0xrxL** [comment]
@ameshkov Where is the new nightly with updated CoreLibs? 🤔

**Comment by jordansworld** [comment]
`from what i understand the fix is in 4.8 nightly. you will have to wait before they release the version with the fix.`

**Comment by sk1llsh0t** [comment]
According to this (https://adguard.com/en/versions/android/nightly.html#version-483) it looks like the first nightly build is somewhere...

**Comment by jordansworld** [comment]
Nightly 1 doesn't include the fix.

**Comment by 0xrxL** [comment]
@derKief @jordansworld @sk1llsh0t Nightly 3 with the new fixed CoreLibs is available. Try it. 👍

**Comment by sk1llsh0t** [comment]
Unfortunately for me the nightly 3 build did not fix the problem.  It still consumes 14%+ cpu and 20% battery consumption.

**Comment by ameshkov** [comment]
@derKief thanks for your help! I understand that it is annoying and we're trying to figure it out.

> Some 4.7 nightlies that I tested did it better. I don't understand it and hope that a solution will finally be found.

The bad thing here is that I don't understand it either.

We stopped at the very same point - we identified the place where the TCP packet is getting stuck, but so far no one was able to send us a log file where the issue was reproduced.

The **only** difference between stable and nightly is having some extra logging. 

Folks, I have to say it again: **we are not able to reproduce it on our devices, we need debug-level logs to figure it out. Ideally, from the latest nightly version.**

At this point the procedure is rather simple:
1. Enable debug-level logging
2. Wait until the issue happens again
3. Export logs, send them to `devteam@adguard.com`

**Comment by derKief** [comment]
@ameshkov I realize that you are trying to solve the problem, but this constant back and forth is not nice and takes time.

I've now tried another test and installed 4.8 nightly 3. So far everything is fine and there are no problems. The last time I installed stable 4.7, the problem was immediately apparent. So I can't understand why the only difference between stable 4.7 and 4.8 nightly 3 is the extended logging.

I'll leave it like this for now and continue to observe.

**Comment by ameshkov** [comment]
@derKief this difference makes me believe that we're dealing with a [race condition](https://en.wikipedia.org/wiki/Race_condition) that is very sensitive to timings, the worst kind. Adding extra logging to these functions adds a microsecond-delay and this improves the situation and makes the issue happen less often. It does not mean that the issue is fully fixed, it should manifest itself eventually.

Some people in comments were saying that they're having that issue on the nightly version too, maybe they can help.

@sk1llsh0t @jordansworld if you can reproduce this with the nightly build, please try collecting the debug log.

**Comment by ameshkov** [comment]
@sk1llsh0t it'd be better if you export the full log via the app and then send an email to `devteam@adguard.com`. Please mention that it's related to this Github issue.

**Comment by sk1llsh0t** [comment]
Logs sent.

**Comment by derKief** [comment]
Short interim report with 4.8 nightly 3: 

- no problems with high load or high frequency on the CPU cores
- no problems with high temperatures. 
- The battery consumption is still too high though. In the Adguard statistics you can clearly see that the background usage is much too high.

I'll let it run until tomorrow but it seems that the problem still exists, although not as extreme as with the stable versions 4.6 and 4.7

**Comment by jordansworld** [comment]
got the best battery usage ive had in a while. 8hr 15min at 20 perceng. 
![Screenshot_20241206_190100_Device care](https://github.com/user-attachments/assets/4c325a99-e4b7-48e9-bc3b-c7d7a93ce91d)
![Screenshot_20241206_190138_Device care](https://github.com/user-attachments/assets/c87e007a-0633-472d-9314-759664657fd8)

**Comment by derKief** [comment]
Final conclusion: 
in direct comparison with my S24, which is still running the stable version 4.5, the A25 with 4.8 nightly 3 consumes twice as much battery (based on the Adguard in-app statistics). And that despite the fact that the S24 is used much more.
In my case, the high battery consumption remains.
Back to Stable 4.5

**Comment by 0x4a4700** [comment]
I have the same issues (high battery drain) with my Galaxy S23 Ultra (latest updates applied).
It started with 4.6, was already in contact with support (
#980394) and thought it was fixed during some nightly I tested, but then upgraded to more recent nightlies and the issues came back with 4.8 build 3 :(
I reverted back now to 4.7 stable and my phone nearly died over night in flight mode because of battery drain of Adguard. 

Where can I download specific nightly apk versions? 

I need especially version 4.7 nightly 48 or later 53 which seem to have worked for me (later on, especially with 4.8 nightly it started again)

**Comment by derKief** [comment]
Any news here ? Anyone tried new stable 4.7.1 ? 

Changelog is somehow not very informative:
_- In this update we've improved the stability of the app and fixed some minor bugs._

And what about 4.8 Nightly 4 ?

**Comment by sk1llsh0t** [comment]
Nightly 4 has the same issue

**Comment by slythrn** [comment]
> Any news here ? Anyone tried new stable 4.7.1 ?
> 
> Changelog is somehow not very informative: _- In this update we've improved the stability of the app and fixed some minor bugs._
> 
> And what about 4.8 Nightly 4 ?

I'm still having issues on my Motorola and Pixel phones with the latest stable version.

**Comment by solkarnar** [comment]
@ameshkov Any updates sir?

**Comment by derKief** [comment]
Has this topic died down or why has there been no update or statement for days?

**Comment by solkarnar** [comment]
> Has this topic died down or why has there been no update or statement for days?

probably holidays and people being away with their families.

**Comment by derKief** [comment]
pretty sure ... but a short status update would have been nice

**Comment by ameshkov** [comment]
Sorry, yeah, holidays indeed mess with us.

Merry Christmas to everyone!

@sk1llsh0t thanks for sending the logs! I may be wrong, but from the log it looks like you have quite a lot of statistics. Am I right? This looks like a legit issue to me, but again it's unrelated to the one I saw in @derKief's logs.

I'll return tomorrow with some specifics.

**Comment by ameshkov** [comment]
@sk1llsh0t confirmed the issue. Temporary solution: clear statistics, the situation will normalize. Proper solution will follow.

**Comment by sk1llsh0t** [comment]
When I tap on the stats tab, I just see 3 green dots in the middle of the screen and nothing loads. The app appears to freeze. I'm not sure how to clear stats without going to that tab. Maybe I'll just wait for the fix to be released.

**Comment by ameshkov** [comment]
@sk1llsh0t you may need to wait for quite some time until the stats are loaded or clear the app data.

Here's the explanation for the issue: AdGuard at startup goes through statistics and tries to "compress" it, i.e. merge records for the old dates. There's a bug in this algorithm that may lead to the issue that you're having.

**Comment by ameshkov** [comment]
@derKief no update unfortunately.

So far your's are the only case where we saw that issue and unlike @sk1llsh0t's issue we cannot reproduce it. With @sk1llsh0t's we've been able to "generate" statistics database that causes the issue, but with this TCP issue we failed despite trying different ways to overload the TCP stack.

What makes it really problematic, it seems that enabling debug-level logging resolves your issue. It may sound strange, but it sometimes happens, it's enough to introduce a slightest delay.

Because of that we'll have to resort to ugly solutions: instead of asking you to collect debug-level log we'll try to detect the problematic condition and crash the app when it happens (and at that point we'll log everything required to troubleshoot it further). This change will come in one of the following nightlies, I'll announce it here.

One more thing: thank you so much to everyone that's helping us figure this all out, I don't think we could manage without you!

**Comment by 0x4a4700** [comment]
I also still have the issue in the latest nightly, with nearly 30% battery usage, also over night when in flight mode... I have askes three weeks ago if there are any apks for previous nightly builds?

**Comment by ameshkov** [comment]
@0x4a4700 which one do you need?

Could you please collect log as described here?
https://github.com/AdguardTeam/AdguardForAndroid/issues/5505#issuecomment-2471045627

**Comment by sodkfjoa** [comment]
I am also getting massive battery drain from Adguard, adb shell top shows 200%+ cpu usage from Adguard and phone is hot. If i turn Adguard off it goes away.

Adguards own battery usage shows 0.00 mAh battery usage, so it seems that the battery usage stats don't work, since adb shell top shows it actually using a lot of cpu.

![image](https://github.com/user-attachments/assets/a56a03bf-a2e9-4299-8eea-ff8d97fe61c8)

**Comment by slythrn** [comment]
When is this going to be fixed? It's been months already.

**Comment by ameshkov** [comment]
We are working on the fix for the statistics issue that was identified a couple of days ago in comments. The change required is rather substantial so it will take more than a week to implement.

@sodkfjoa in your case when you observe 100% cpu what thread is taking all that?

Run this command to see that:
`adb shell "top -H -p PROCESS_ID"`

If this is a "statistics" thread then this is it.

**Comment by 0x4a4700** [comment]
@ameshkov 
I have analysed this further with ADB and I see the following issues, at least for my phone/usage. During the night I put the phone in flight mode and Adguard is constantly looping/trying to access the Internet (DNS server) which seems to take a really high hit (the specific Thread in the logs is the one with high load in the output of top). I am not sure if this was the same in older versions, but this is taking really high CPU load for me now. 

I am using the Stremio streaming app and I think it could also be one of the issues during the day (maybe also during the night, if the server process constantly tries to connect to some trackers). While streaming, Adguard will take about 15-25% CPU in the tcpip-stack-run & tcpip-stack-event threads according to the "top" output.
I rarely see the statistics-pool taking high CPU, in comparison to the other threads, it does not take that much CPU (at least not constantly, only saw 100% CPU once there, so this one is not the issue for me I guess).

I have sent the logs, top output and screenshots to devteam@adguard.com

**Comment by ameshkov** [comment]
@0x4a4700 thanks a lot for collecting all these logs, this is very helpful!

Your issue is different and it's indeed caused by Streamio, but we can and should mitigate this on our side.
The explanation is here: https://github.com/AdguardTeam/AdguardForAndroid/issues/5571

Briefly, when the device does not have connection to the Internet, Streamio still thinks that the device is connected and continues to send DNS queries. AdGuard emulates a "SERVFAIL" response since it cannot resolve these queries without Internet connection, but Streamio just re-sends the DNS query again and therefore we get into an infinite loop of failed DNS requests. Temporary solution: disable filtering for Streamio via AdGuard's app management. Regarding day-time usage, I don't see any issues there, consuming 15-20% on filtering a "heavy" connection is okay.

**Comment by 0x4a4700** [comment]
thanks very much! Stremio had already disabled filtering, but I have also disabled the traffic routing completely now.

**Comment by derKief** [comment]
Any news?  Will this ever be resolved?
Version 4.5 is getting old.

**Comment by artemiv4nov** [comment]
Next week we will provide a new solution by the nightly track, I will notify about it later.

**Comment by adrienrichmond** [comment]
Hey, I was using Adguard, the latest of the name, my battery lost 10% of its capacity over a whole day whereas before it was only 1 to 2%. Do you have a solution in the meantime? Thank you

**Comment by derKief** [comment]
another week ...  no solution :-(

**Comment by artemiv4nov** [comment]
We are currently experiencing some issues with the release of the Nightly version due to large code changes, please wait a few more days.

**Comment by 0xrxL** [comment]
@artemiv4nov There are still problems with release?

**Comment by artemiv4nov** [comment]
We are working up a sweat and realize the importance. I hope to finish this/next week, sorry for another delay, the changes regarding the statistics functionality are really big.

**Comment by 0xrxL** [comment]
> We are working up a sweat and realize the importance. I hope to finish this/next week, sorry for another delay, the changes regarding the statistics functionality are really big.

Hope will not be released too late, because I can systematically reproduce an issue with Google, which show an anti-captcha request due to a "large number of requests sent".

**Comment by derKief** [comment]
any news on this ?

**Comment by artemiv4nov** [comment]
Everything is 95% finished, there will be a new Nightly version next week

**Comment by adrienrichmond** [comment]
it's sooo long, battery drain is tiresome

**Comment by 0xrxL** [comment]
> it's sooo long, battery drain is tiresome

They released a new nightly today, but it's a weird version, because is not marked as nightly despite is available on nightly channel. Try it in any case and let us know if everything is fixed.

![Image](https://github.com/user-attachments/assets/4cff82f9-7a1b-4590-9975-01d3d5d6b104)

**Comment by artemiv4nov** [comment]
This version doesn't yet contain the needed battery changes, though it has a huge chunk of changed code. I will notify you separately when a version with battery changes will be available in the Nightly track

**Comment by 0xrxL** [comment]
@artemiv4nov I can understand the difficult to merge a huge amount of new code and make a new version available, but 3 weeks are already passed and the only working alternative is AdGuard DNS, which unfortunately doesn't apply cosmetics changes. I can only say that the web navigation, with empty spaces here and there (including privacy popups too), is terrible.

**Comment by jordansworld** [comment]
> This version doesn't yet contain the needed battery changes, though it has a huge chunk of changed code. I will notify you separately when a version with battery changes will be available in the Nightly track

can you notify me when the version is out that fixes the battery

**Comment by artemiv4nov** [comment]
So, Nightly 7 is released, the problematic statistic functionality has been replaced, let's test this version @jordansworld @0xrxL @adrienrichmond @derKief

**Comment by jordansworld** [comment]
i went to update but dont see it. @artemiv4nov

**Comment by adrienrichmond** [comment]
> So, Nightly 7 is released, the problematic statistic functionality has been replaced, let's test this version @jordansworld @0xrxL @adrienrichmond @derKief 

You have link please ?

**Comment by 0xrxL** [comment]
@jordansworld @adrienrichmond you can download it from public beta/nightly:

https://adguard.com/it/beta.html

**Comment by jordansworld** [comment]
it finally shows up in the app thankyou

**Comment by 0xrxL** [comment]
@jordansworld check if you can reproduce a stuck page loading too, together with abnormal battery consuption.

I was able to reproduce both on previous nightlies.

**Comment by jordansworld** [comment]
will let you know about the battery. as far as the stuck page loading i think they fixed it but havent merged in that code:

https://github.com/AdguardTeam/AdguardForAndroid/issues/5596

**Comment by 0xrxL** [comment]
@artemiv4nov DNSLibs will be pushed this week too? Stuck loading issue is another annoying thing.

**Comment by adrienrichmond** [comment]
Fix working ?

**Comment by jordansworld** [comment]
so the battery seems to be better.

**Comment by slythrn** [comment]
The latest nightly is making my phone very warm. (Motorola ThinkPhone)

**Comment by derKief** [comment]
tried Nightly 8 but unfortunately still higher battery consumption compared to Stable 4.5

I'm tired of it now and have therefore uninstalled the Adguard app on all my smartphones. will now be set to pure DNS filtering. Let's see how I get on with it. 
I may give the app another try at a later date.

**Comment by slythrn** [comment]
I also uninstalled it from all the phones in my household. I can't deal with the battery drain and excessive heating issues anymore.

**Comment by jordansworld** [comment]
I have to be the only one without battery issues. Lol. Is it because I'm on Samsung? My s23 didn't have the issue on the newest and either does my s25

**Comment by Nicos18** [comment]
The same as @jordansworld said in the previous post.

With S23 no battery issues.

Latest stable version 4.7.1

![Image](https://github.com/user-attachments/assets/4f75d3ae-23cf-4081-a4d4-c8a4331d7dcf)

**Comment by derKief** [comment]
Statistics bug maybe fixed but

**One issue is due to a race condition in the TCP/IP stack implement. VERY random bug, hard to catch.**

seems to be NOT fixed.

**Comment by 0xrxL** [comment]
NOTE: if any of you can reproduce the "infinite loading" issue, just disable "legitimate URL reduction" filters list to fix it.

P.S: An issue is currently opened on the github of this filters list.

**Comment by war59312** [comment]
Just got hit with this on 4.8 beta 1 on my galaxy 24+. Was so bad phone not even charging when using super charger at 45w. Back to 4.7 and things seem fine.

**Comment by ameshkov** [comment]
The current v4.8 nightly and beta channels contain two important changes:

1. The stats bug, that's the one that was causing most of the issues for people in the thread. Before making any assumptions give it a go for a couple of hours and then check AdGuard's battery stats. We are monitoring the results now on our test devices and monitoring the feedback.

2. @derKief's TCP/IP issue: we added a workaround that should prevent the infinite loop and (IMPORTANT) when it runs into this "infinite loop" issue, it prints additional information to the debug log. If you're able to spare time on one more troubleshooting attempt, I'll be very grateful for this.

Now regarding other reports in this issue.

@jordansworld @Nicos18 I am glad to hear that you have no issues. To be honest, this seems to be a norm for the majority of users as running into everything we discovered here required additional conditions.

@0xrxL your report has nothing to do with this issue or with battery issues in general, let's discuss it in the relevant thread. At this moment I'm waiting for additional report and I'll instruct QA on what they should pay attention to.

@slythrn @war59312 it'd be very helpful if you could collect logs and thread dumps as described here in this comment: https://github.com/AdguardTeam/AdguardForAndroid/issues/5505#issuecomment-2471045627
I promise I'll be able to very quickly troubleshoot the issue if you do it.

**Comment by 0xrxL** [comment]
@ameshkov I removed the useless comments and tagged you in the right thread.

**Comment by 0xrxL** [comment]
@ameshkov seems that a user found a way to fix another app behaviour that consume battery: https://github.com/AdguardTeam/AdguardForAndroid/issues/5599

**Comment by JSCFORCE** [comment]
I updated to v4.8 Nightly 9 and it's still draining my battery.

When I attempt to use the "report a bug" feature it won't send unless I uncheck the box "send logs and system info"

I assume that makes the report less useful to you.

**Comment by jordansworld** [comment]
here is my third day on my new galaxy s25 ultra. latest nightly.

![Image](https://github.com/user-attachments/assets/a7942f01-85b3-4ac2-aec4-6d49bf2b335a)
![Image](https://github.com/user-attachments/assets/cf8da3c6-2b5e-4fe3-b3e7-4c451c3413b1)

**Comment by JSCFORCE** [comment]
I updated to v4.8 Nightly 11 and it's still draining my battery.

The app was using 6.6GB of cache.

So I cleared that and resent a bug report with the logs attached after the battery was still draining while it was charging.

**Comment by ameshkov** [comment]
@JSCFORCE where exactly have you sent the report? Can you tell me the ticket ID?

**Comment by JSCFORCE** [comment]
> [@JSCFORCE](https://github.com/JSCFORCE) where exactly have you sent the report? Can you tell me the ticket ID?

I submitted it through the app itself.

**Comment by ameshkov** [comment]
@JSCFORCE

Submitting through the app is only enough for troubleshooting simple issues, this one is hardly it.
But anyway, if you submitted through the app, you should've received an autoreply with the ticket ID.

If this is easy to reproduce this issue, please do the following:

1. Change the logging level to "Record everything" in the app's advanced settings.
2. Reproduce the issue
3. Export the logs and send them to `devteam@adguard.com`, mention your Github username and this Github issue in the email.

**Comment by JSCFORCE** [comment]
@ameshkov 

I have now done so. Thank you.

**Comment by ameshkov** [comment]
@JSCFORCE the nightly update should be available in 5-10 minutes.

Please use AG for 1-2 hours and check if the issue happens again.

**Comment by JSCFORCE** [comment]
@ameshkov 

I will give it a shot and let you know. Thank you very much.

**Comment by ameshkov** [comment]
@JSCFORCE how it goes? We were thinking about shipping it to the beta channel, but waiting for you to confirm that the issue is gone.

**Comment by JSCFORCE** [comment]
@ameshkov 

sorry bro. was at work all day.

we are a go. used it all day on and off charger. 

been charging normal.

good job.

**Comment by jordansworld** [comment]
anybody elses battery greatly improved? on the 7th i got my s25 ultra. was getting 7 hours of sot. i now get 10

**Comment by adrienrichmond** [comment]
Bug fixed Ad guard 4.8 ?

**Comment by jordansworld** [comment]
> Bug fixed Ad guard 4.8 ?

yes i think so. but latest version is 4.9

**Comment by 0xrxL** [comment]
> > Bug fixed Ad guard 4.8 ?
> 
> yes i think so. but latest version is 4.9

He was referring to 4.8 stable released today, which contains all changes from 4.9 Nightly 1.

**Comment by 0xrxL** [comment]
@derKief Try last 4.8 stable, maybe this issue can be finally closed. 👍

**Comment by ameshkov** [comment]
I intend to close this issue if we don't receive any new reports with logs.

If you're experiencing any issue like this one, please collect logs [as described here](https://github.com/AdguardTeam/AdguardForAndroid/issues/5505#issuecomment-2668864238) and send them to `devteam@adguard.com`, we'll either reopen this one or open a new issue.

**Comment by derKief** [comment]
still testing new stable 4.8.
For now it looks good so far but this is no final for now. Will wait running charging cycle to have a comparison. Will report back.

**Comment by ameshkov** [comment]
@derKief first of all, great news and thank you for testing it!

However, I'd like to ask you to collect "record everything" logs for a couple hours of work **even if everything looks okay**.

The reason for this is that even though we added some code to handle your specific issue (TCP stack bug), we're still very interested in learning the actual root cause of it and we added additional logging for that.

**Comment by jordansworld** [comment]
i can do some logging for you too @ameshkov so just use my phone as i would normally but with debug logging on?

**Comment by jordansworld** [comment]
im curious to see if different phone brands have different issues?

**Comment by ameshkov** [comment]
@jordansworld the only person for whom I saw this TCP bug happening is @derKief, everyone else in this thread was suffering from a different one :)

We created quite a bit of confusion by solving both of them in one issue, I should've filed a different one.

**Comment by derKief** [comment]
@ameshkov
atm i'm testing on my main device an Samsung S25 ... as mentioned that looks promising at the moment.
The device you are talking about (TCP bug) is my second device an Samsung A25.
I don't have Adguard installed on it atm. I want to wait for the test on the S25 first. 
One thing at a time ;-)

**Comment by ameshkov** [comment]
@jordansworld please, pass the logs via email (`devteam@adguard.com`), debug-level logging is not supposed to be shared on Github.

**Comment by jordansworld** [comment]
@ameshkov sent. The file is 95mb so i had to upload via a file host

**Comment by derKief** [comment]
@ameshkov 
i installed Stable 4.8 also on my Samsung A25. The extended logging was activated  ~ last 3 hours. The logs just sent.
It seems that A25 is also running fine for now. 
Will continue testing.

**Comment by 0xrxL** [comment]
@derKief Try to use the app in foreground (for example: go on statistics and/or into settings) for 10 minutes at least, and check if the battery consuption will increase over 1%. This could be another unfixed power drain.

**Comment by jordansworld** [comment]
were you guys able to check my logs? @0xrxL

**Comment by 0xrxL** [comment]
> were you guys able to check my logs? [@0xrxL](https://github.com/0xrxL)

I'm not an adguard dev 😅

Wait a response from @ameshkov

**Comment by jordansworld** [comment]
> > were you guys able to check my logs? [@0xrxL](https://github.com/0xrxL)
> 
> I'm not an adguard dev 😅
> 
> Wait a response from [@ameshkov](https://github.com/ameshkov)

ohh 🤣🤣🤣

**Comment by derKief** [comment]
A short interim report. 
The A25 does use more battery power compared to the s25, even though it is not used as much as the s25. 
In particular, the CPU background usage on the A25 is much higher. Checked and compared with the in-app statistics.
So for now I would currently say:
S25 is good.
A25 is not really good resp. could be better.

@ameshkov did you already analyze the A25 logs i sent you? Any abnormalities ?

**Comment by jordansworld** [comment]
i think the app uses alot of battery when ur using it in the foreground.

![Image](https://github.com/user-attachments/assets/b0b32a27-338b-4e6f-8842-5df4eca6d4fe)

**Comment by derKief** [comment]
The main difference in the use of my two devices is that I almost only use the native apps on the S25, but do almost everything in the browser on the A25. But that doesn't explain the increased CPU background usage on the A25 compared to the S25.

**Comment by derKief** [comment]
> Actually might! Browser is more intensively filterable than average app, so presumably uses more CPU from AG.
> 
I am aware of that ... but if you look closely I wrote about background CPU not foreground CPU


> Maybe try more apples-to-apples via shifting usage pattern on 1 device to match other's & see whether stats reflect that?
>
I agree with you... but I have no desire to adjust everything.

**Comment by TPS** [comment]
> I am aware of that ... but if you look closely I wrote about background CPU not foreground CPU

Both do filtering (I'm unaware _precisely_ which filtering is separated into foreground/background 🤷🏾‍♂️), so Idk why that's relevant?

**Comment by ameshkov** [comment]
@derKief hi, thanks for the logs!

I've skimmed to it and generally the usage does not seem too high, not even close to what it was before. According to the log the total CPU usage for the day (starting at 8am to 9pm) seems to be about **6.5-7 minutes** of CPU time which is completely fine.

The heaviest period was between 18:19 and 18:42 when AdGuard used about 70 seconds of CPU time.
In that period AdGuard has filtered 1338 HTTP requests and 702 DNS requests.

> background CPU not foreground CPU

Foreground means CPU usage when AdGuard UI is open, i.e. when you use the app UI.
Rendering UI is generally much heavier than doing the filtering itself, that's why there's a distinction in our stats.

> But that doesn't explain the increased CPU background usage on the A25 compared to the S25.

Do you compare CPU usage minutes or mAh? S25 has 8 CPU cores and A25 has 4 CPU cores so CPU minutes are translated to a twice smaller usage on S25. For better understanding, CPU time is roughly equal to the total time a single CPU core was running with 100% load.

Also, @TPS is right about filtering browser being a more complicated task than everything else as it includes HTTPS filtering. 

All in all, I don't see any issues with it in the log.

The problem is that I don't see the original TCP bug there either. My only assumption is that it was somehow triggered by the increased usage caused by the statistics module and now since it's fixed it's not triggered anymore. Well, better than nothing.

@jordansworld I checked your log, it's also fine, thank you!

**Comment by derKief** [comment]
@ameshkov thanks a lot for checking the logs and the the good explanation.
